In [79]:
import pandas as pd
import numpy as np
df = pd.read_excel("Day35_Pandas_DataCleaning.xlsx", sheet_name = 'Raw Data', header=1)
print(df.shape)
print(df.dtypes)

(20, 10)
order_id           int64
customer_name     object
email             object
city              object
category          object
quantity         float64
unit_price         int64
discount         float64
order_date        object
status            object
dtype: object


In [80]:
# Check nulls
print(df.isnull().sum())
print(df.isnull().sum()/len(df)*100) #as%

order_id         0
customer_name    0
email            1
city             0
category         0
quantity         2
unit_price       0
discount         5
order_date       0
status           0
dtype: int64
order_id          0.0
customer_name     0.0
email             5.0
city              0.0
category          0.0
quantity         10.0
unit_price        0.0
discount         25.0
order_date        0.0
status            0.0
dtype: float64


In [81]:
# Check duplicates
print(df.duplicated().sum())
print(df[df.duplicated(keep=False)]) # show all duplicate rows

0
Empty DataFrame
Columns: [order_id, customer_name, email, city, category, quantity, unit_price, discount, order_date, status]
Index: []


In [82]:
df['discount']=df['discount'].fillna(0.0)
print(df['discount'])

0     0.10
1     0.00
2     0.05
3     0.10
4     0.00
5     0.00
6     0.00
7     0.15
8     0.00
9     0.00
10    0.05
11    0.10
12    0.08
13    0.00
14    0.10
15    0.00
16    0.12
17    0.05
18    0.00
19    0.00
Name: discount, dtype: float64


In [83]:
df['quantity']=df['quantity'].fillna(df['quantity'].median())
print(df['quantity'])

0     2.0
1     1.0
2     2.0
3     1.0
4     3.0
5     2.0
6     1.0
7     4.0
8     2.0
9     1.0
10    2.0
11    2.0
12    3.0
13    1.0
14    2.0
15    5.0
16    2.0
17    1.0
18    3.0
19    2.0
Name: quantity, dtype: float64


In [108]:
df = df[df['email'].str.strip() !=""]
print(df['email'])

0          arjun@gmail.com
1     priyasingh@gmail.com
2          rahul@yahoo.com
3          arjun@gmail.com
4        sneha@outlook.com
5         vikram@gmail.com
6        deepika@gmail.com
7           amit@gmail.com
8               neha.gupta
9         rajesh@gmail.com
10         meena@gmail.com
11        suresh@gmail.com
12         kavya@gmail.com
13          nita@yahoo.com
14         pooja@gmail.com
15          ravi@gmail.com
16       lakshmi@gmail.com
Name: email, dtype: object


In [85]:
assert df['discount'].isnull().sum() == 0 
print(df.isnull().sum())

order_id         0
customer_name    0
email            0
city             0
category         0
quantity         0
unit_price       0
discount         0
order_date       0
status           0
dtype: int64


#### The discount column has the biggest percentage of missing values at 25%(5 out of 20 rows).Leaving these nulls unaddressed created a business problem because it prevents accurate financial calculations such as total revenue. Additionally missing a quarter of data for this field makes it impossible to reliably analyse the sales leading to incomplete information.Additionally there is a secondary issue i the email column, while 5% of the records are missing.While the discount nulls hinder financial analysis. the missing email represents a different data gap a missing identifier that prevents order confirmation and directly hinders customer communication.


In [86]:
print(df[df.duplicated(subset=['customer_name','email','category','unit_price']
                    ,keep=False)])


    order_id customer_name                 email    city     category  \
0       1001  Arjun Sharma       arjun@gmail.com  mumbai  Electronics   
1       1002   Priya Singh  priyasingh@gmail.com   Delhi     Clothing   
6       1007   Priya Singh  priyasingh@gmail.com   delhi     Clothing   
14      1015  Arjun Sharma       arjun@gmail.com  Mumbai  Electronics   

    quantity  unit_price  discount  order_date     status  
0        2.0       15000       0.1  2024-01-15  Delivered  
1        1.0        2500       0.0  2024-01-18  Delivered  
6        1.0        2500       0.0  2024-03-05  Delivered  
14       2.0       15000       0.1  2024-05-20  Delivered  


In [87]:
print(df.shape)

(19, 10)


In [88]:
df = df.drop_duplicates(subset=['customer_name','email','category','unit_price']
                    ,keep='first')
print(df)

    order_id customer_name                 email       city        category  \
0       1001  Arjun Sharma       arjun@gmail.com     mumbai     Electronics   
1       1002   Priya Singh  priyasingh@gmail.com      Delhi        Clothing   
2       1003   Rahul Mehta       rahul@yahoo.com  Bangalore     Electronics   
3       1004  arjun sharma       arjun@gmail.com     Mumbai     Electronics   
4       1005   Sneha Patel     sneha@outlook.com  Ahmedabad  Home & Kitchen   
5       1006    Vikram Rao      vikram@gmail.com    Chennai        Clothing   
7       1008  Deepika Nair     deepika@gmail.com      Kochi  Home & Kitchen   
8       1009    Amit Kumar        amit@gmail.com       Pune     Electronics   
9       1010    Neha Gupta            neha.gupta    Lucknow        Clothing   
10      1011  Rajesh Verma      rajesh@gmail.com     Jaipur  Home & Kitchen   
11      1012     Meena Das       meena@gmail.com    Kolkata     Electronics   
12      1013  Suresh Kumar      suresh@gmail.com  Hy

In [89]:
print(df.shape)

(17, 10)


In [90]:
df.reset_index(drop = True, inplace=True)
print(df)

    order_id customer_name                 email       city        category  \
0       1001  Arjun Sharma       arjun@gmail.com     mumbai     Electronics   
1       1002   Priya Singh  priyasingh@gmail.com      Delhi        Clothing   
2       1003   Rahul Mehta       rahul@yahoo.com  Bangalore     Electronics   
3       1004  arjun sharma       arjun@gmail.com     Mumbai     Electronics   
4       1005   Sneha Patel     sneha@outlook.com  Ahmedabad  Home & Kitchen   
5       1006    Vikram Rao      vikram@gmail.com    Chennai        Clothing   
6       1008  Deepika Nair     deepika@gmail.com      Kochi  Home & Kitchen   
7       1009    Amit Kumar        amit@gmail.com       Pune     Electronics   
8       1010    Neha Gupta            neha.gupta    Lucknow        Clothing   
9       1011  Rajesh Verma      rajesh@gmail.com     Jaipur  Home & Kitchen   
10      1012     Meena Das       meena@gmail.com    Kolkata     Electronics   
11      1013  Suresh Kumar      suresh@gmail.com  Hy

#### Using order_id alone fails because systems often generate a unique ID for every click, meaning a duplicate transaction would still have two different IDs. In a business a duplicate order is redundent record of the same purchase identifying them prevents shipping errors and double billing the customer.

In [91]:
df['city']=df['city'].str.strip().str.title()
print(df['city'].unique())

['Mumbai' 'Delhi' 'Bangalore' 'Ahmedabad' 'Chennai' 'Kochi' 'Pune'
 'Lucknow' 'Jaipur' 'Kolkata' 'Hyderabad' 'Surat']


In [92]:
df['customer_name'] = df['customer_name'].str.strip().str.title()
print(df['customer_name'].unique())

['Arjun Sharma' 'Priya Singh' 'Rahul Mehta' 'Sneha Patel' 'Vikram Rao'
 'Deepika Nair' 'Amit Kumar' 'Neha Gupta' 'Rajesh Verma' 'Meena Das'
 'Suresh Kumar' 'Kavya Reddy' 'Nita Shah' 'Pooja Joshi' 'Ravi Chandra'
 'Lakshmi Iyer']


In [93]:
df['valid_email'] = df['email'].str.contains('@')
invalid_count = (df['valid_email']== False).sum()
print(invalid_count)

1


In [94]:
df['status'] = df['status'].str.strip().str.title()
print(df['status'].value_counts())

status
Delivered    12
Cancelled     2
Pending       2
Shipped       1
Name: count, dtype: int64


#### Pandas is a case sensitive so it treats 'mumbai', 'Mumbai', 'MUMBAI' as three distinct groups. For example if you use .groupby('city').sum(), instead of seeing a total for Mumbai you see three seperate rows with split totals, making your report inaccurate.

In [95]:
df['order_date'] = pd.to_datetime(df['order_date'])
df['order_month']=df['order_date'].dt.month
print(df['order_date'])
print(df['order_month'])

0    2024-01-15
1    2024-01-18
2    2024-02-02
3    2024-02-10
4    2024-02-14
5    2024-03-01
6    2024-03-12
7    2024-04-01
8    2024-04-08
9    2024-04-15
10   2024-05-01
11   2024-05-10
12   2024-05-18
13   2024-06-03
14   2024-06-20
15   2024-07-04
16   2024-07-14
Name: order_date, dtype: datetime64[ns]
0     1
1     1
2     2
3     2
4     2
5     3
6     3
7     4
8     4
9     4
10    5
11    5
12    5
13    6
14    6
15    7
16    7
Name: order_month, dtype: int32


In [96]:
df['revenue'] = (df['quantity'] * df['unit_price']*(1-df['discount'])).round(2)
print(df['revenue'])

0     27000.0
1      2500.0
2     17098.1
3     13500.0
4      3600.0
5      7000.0
6      2720.0
7     24000.0
8      1800.0
9      1805.0
10     9900.0
11     6072.0
12    22000.0
13     3000.0
14     1425.0
15    22500.0
16     4000.0
Name: revenue, dtype: float64


In [97]:
df['quantity'] = df['quantity'].astype(int)
print(df['quantity'])

0     2
1     1
2     2
3     1
4     3
5     2
6     4
7     2
8     1
9     2
10    2
11    3
12    1
13    5
14    1
15    3
16    2
Name: quantity, dtype: int64


In [98]:
assert df['order_date'].dtype == 'datetime64[ns]'
assert df['quantity'].dtype == 'int64'
assert df['discount'].dtype == 'float64'
assert df['valid_email'].dtype == 'bool'
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17 entries, 0 to 16
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   order_id       17 non-null     int64         
 1   customer_name  17 non-null     object        
 2   email          17 non-null     object        
 3   city           17 non-null     object        
 4   category       17 non-null     object        
 5   quantity       17 non-null     int64         
 6   unit_price     17 non-null     int64         
 7   discount       17 non-null     float64       
 8   order_date     17 non-null     datetime64[ns]
 9   status         17 non-null     object        
 10  valid_email    17 non-null     bool          
 11  order_month    17 non-null     int32         
 12  revenue        17 non-null     float64       
dtypes: bool(1), datetime64[ns](1), float64(2), int32(1), int64(3), object(5)
memory usage: 1.7+ KB


In [102]:
assert df.isnull().sum().sum() == 0
print(df.isnull().sum())

order_id         0
customer_name    0
email            0
city             0
category         0
quantity         0
unit_price       0
discount         0
order_date       0
status           0
valid_email      0
order_month      0
revenue          0
dtype: int64


In [107]:
top_cities = df.groupby('city')['revenue'].sum().nlargest(3)
print(top_cities)

city
Bangalore    43098.1
Mumbai       40500.0
Chennai      29500.0
Name: revenue, dtype: float64


#### Bangalore - ₹43,098.1: The primary revenue driver.Increase inventory for top-selling categories in this hub.
#### Mumbai - ₹40,500.6.A very strong second.Run targeted marketing campaigns to close gap with Bangalore.
#### Chennai - ₹29,500.0:Solid regional contributor.Introduce loyalty rewards to increase the average order value.

#### The data contains three critical issues: null discounts in 5 rows lead to incorrect revenue calculations.Action taken to apply.fillna(0) to treat missing values to zero.
#### While 4 duplicate rows consisting of 2 pairs(Arjun at rows 0 and 14, and Priya at rows 1 and 6) falsely inflate both order counts and total sales.Action take was to use .drop_duplicates() to ensure each transaction was unique.
#### Additionally inconsistent city casing(e.g.,'Mumbai' vs 'mumbai') fragments reporting by splitting single cities into multiple groups during analysis.Action taken was to standardize the column using .str.strip(),str.title() to provide a unified view of performance.
#### Addressing these errors is essential for ensuring financial accuracy and reliable business insights.